core/src/crds_gossip_pull.rs

```diff
@@ -542,20 +542,12 @@ impl CrdsGossipPull {
         now: u64,
         timeouts: &HashMap<Pubkey, u64>,
     ) -> usize {
-        let old = crds.find_old_labels(now, timeouts);
```

Comment:
I wonder what the time separation is here, because we could more easily split the work into a read-lock/write-lock split here. `purged_values` could be something like `Vec<Vec<_>>` and just appending a `Vec` onto the end would be super cheap within the write lock part. You also have `crds.remove` which doesn't seem easy to refactor outside the write lock, so if that is the main time then maybe it doesn't make sense. But you don't have to do a full table scan within the write lock.

Reply:
yeah this was done given that we already have a mutable reference here, so might implement the same logic without the extra heap allocation, cloning and look-ups.
From profiling run-time of a small gce network, it seems like vast majority of the time is spent in scanning the table. if that is also the case on mainnet, then yes you are right, better not to do a full table scan holding the write-lock.
From the metrics seems like the purge-count tends to be very small in general. So again the table scan part should be dominant. Though perhaps maybe one node is having large `purge-count` compared to rest of the nodes (which seems odd), and that would involve different trade-offs.

Alternatives are implement a finer grained locking in gossip (and perhaps one dedicated read/write lock for crds table), or utilize this thread_pool:
https://github.com/solana-labs/solana/blob/87de82ac9/core/src/cluster_info.rs#L1562
in the scan part of handle_purge further down:
https://github.com/solana-labs/solana/blob/87de82ac9/core/src/cluster_info.rs#L1606
![cluster_info-purge-count-tail](https://user-images.githubusercontent.com/1288998/95260229-e5f15300-0817-11eb-876a-db451629fe17.png)